File: isegoria/src/main/java/com/eulersbridge/isegoria/profile/badges/BadgeAdapter.java
package com.eulersbridge.isegoria.profile.badges;

import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.RequestManager;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.API;
import com.eulersbridge.isegoria.network.api.models.Badge;
import com.eulersbridge.isegoria.network.api.responses.PhotosResponse;
import com.eulersbridge.isegoria.util.network.SimpleCallback;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Response;

class BadgeAdapter extends RecyclerView.Adapter<BadgeViewHolder> {

    final private List<Badge> completedItems = new ArrayList<>();
    final private List<Badge> remainingItems = new ArrayList<>();

    final private RequestManager glide;
    final private API api;

    BadgeAdapter(@NonNull RequestManager glide, @NonNull API api) {
        this.glide = glide;
        this.api = api;
    }

    void replaceCompletedItems(@NonNull List<Badge> newItems) {
        completedItems.clear();
        completedItems.addAll(newItems);
        notifyItemRangeChanged(0, newItems.size());
    }

    void replaceRemainingItems(@NonNull List<Badge> newItems) {
        remainingItems.clear();
        remainingItems.addAll(newItems);

        int remainingItemsStartIndex = remainingItems.size() - 1;

        // remainingItemsStartIndex < 0 if remainingItems.size() == 0
        if (remainingItemsStartIndex < 0)
                remainingItemsStartIndex = 0;

        // Remaining items show after completed items
        notifyItemRangeChanged(remainingItemsStartIndex, newItems.size());
    }

    @Override
    public int getItemCount() {
        return completedItems.size() + remainingItems.size();
    }

    private int getImageIndex() {
        switch (Resources.getSystem().getDisplayMetrics().densityDpi) {
            case DisplayMetrics.DENSITY_LOW:
                return 5;
            case DisplayMetrics.DENSITY_MEDIUM:
                return 4;
            default:
                return 3;
        }
    }

    @Override
    public void onBindViewHolder(BadgeViewHolder viewHolder, int index) {

        Badge item;
        boolean completed = false;

        if (index < completedItems.size()) {
            item = completedItems.get(index);
            completed = true;

        } else if (index < remainingItems.size()) {
            item = remainingItems.get(index);

        } else {
            viewHolder.setItem(null, false);
            return;
        }

        viewHolder.setItem(item, completed);

        final int imageIndex = getImageIndex();
        final long itemId = item.id;

        WeakReference<BadgeViewHolder> weakViewHolder = new WeakReference<>(viewHolder);

        api.getPhotos(itemId).enqueue(new SimpleCallback<PhotosResponse>() {
            @Override
            protected void handleResponse(Response<PhotosResponse> response) {
                PhotosResponse body = response.body();

                if (body != null
                        && body.totalPhotos > (imageIndex + 1)) {
                    BadgeViewHolder innerViewHolder = weakViewHolder.get();

                    if (innerViewHolder != null) {
                        String imageUrl = body.photos.get(imageIndex).thumbnailUrl;

                        if (!TextUtils.isEmpty(imageUrl))
                            innerViewHolder.setImageUrl(glide, itemId, imageUrl);
                    }
                }
            }
        });
    }

    @Override
    public BadgeViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).
                inflate(R.layout.profile_badges_list_item, viewGroup, false);
        return new BadgeViewHolder(itemView);
    }

    @Override
    public void onViewRecycled(BadgeViewHolder holder) {
        holder.onRecycled();

        super.onViewRecycled(holder);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/profile/badges/BadgeAdapter.kt
package com.eulersbridge.isegoria.profile.badges

import android.content.res.Resources
import android.support.v7.widget.RecyclerView
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.ViewGroup
import com.bumptech.glide.RequestManager
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.API
import com.eulersbridge.isegoria.network.api.model.Badge
import com.eulersbridge.isegoria.util.extension.subscribeSuccess
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import java.lang.ref.WeakReference
import java.util.*

internal class BadgeAdapter(
    private val glide: RequestManager,
    private val api: API
) : RecyclerView.Adapter<BadgeViewHolder>() {

    private val compositeDisposable = CompositeDisposable()
    private val completedItems = ArrayList<Badge>()
    private val remainingItems = ArrayList<Badge>()

    fun replaceCompletedItems(newItems: List<Badge>) {
        completedItems.clear()
        completedItems.addAll(newItems)
        notifyItemRangeChanged(0, newItems.size)
    }

    fun replaceRemainingItems(newItems: List<Badge>) {
        remainingItems.clear()
        remainingItems.addAll(newItems)

        var remainingItemsStartIndex = remainingItems.size - 1

        // remainingItemsStartIndex < 0 if remainingItems.size() == 0
        if (remainingItemsStartIndex < 0)
            remainingItemsStartIndex = 0

        // Remaining items show after completed items
        notifyItemRangeChanged(remainingItemsStartIndex, newItems.size)
    }

    override fun getItemCount() = completedItems.size + remainingItems.size

    private fun getImageIndex(): Int {
        return when (Resources.getSystem().displayMetrics.densityDpi) {
            DisplayMetrics.DENSITY_LOW -> 5
            DisplayMetrics.DENSITY_MEDIUM -> 4
            else -> 3
        }
    }

    override fun onBindViewHolder(viewHolder: BadgeViewHolder, index: Int) {

        val item: Badge
        var completed = false

        when {
            (index < completedItems.size) -> {
                item = completedItems[index]
                completed = true
            }
            (index < remainingItems.size) -> {
                item = remainingItems[index]
            }
            else -> {
                viewHolder.setItem(null, false)
                return
            }
        }

        viewHolder.setItem(item, completed)

        val imageIndex = getImageIndex()
        val itemId = item.id

        val weakViewHolder = WeakReference(viewHolder)

        api.getPhotos(itemId).subscribeSuccess {
            if (it.totalPhotos > imageIndex + 1) {
                val innerViewHolder = weakViewHolder.get()

                if (innerViewHolder != null)
                    it.photos?.get(imageIndex)?.getPhotoUrl()
                        ?.takeUnless { it.isBlank() }
                        ?.also {
                            innerViewHolder.setImageUrl(glide, item.id, it)
                        }
            }
        }.addTo(compositeDisposable)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): BadgeViewHolder {
        val itemView = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.profile_badges_list_item, viewGroup, false)
        return BadgeViewHolder(itemView)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        compositeDisposable.dispose()
    }
}